src/auth/oauth2common.ts

```diff
@@ -0,0 +1,212 @@
+ * application/x-www-form-urlencoded content types for HTTP methods that support
+ * request bodies are supported.
+ */
+export abstract class OAuthClientAuthHandler {
+  private crypto: Crypto;
+
+  /**
+   * Instantiates an OAuth client authentication handler.
+   * @param clientAuthentication The client auth credentials.
+   */
+  constructor(private readonly clientAuthentication?: ClientAuthentication) {
+    this.crypto = createCrypto();
+  }
+
+  /**
+   * Applies client authentication on the OAuth request's headers or POST
+   * body but does not process the request.
+   * @param opts The GaxiosOptions whose headers or data are to be modified
+   *   depending on the client authentication mechanism to be used.
+   * @param bearerToken The optional bearer token to use for authentication.
+   *   When this is used, no client authentication credentials are needed.
+   */
+  protected applyClientAuthenticationOptions(
+    opts: GaxiosOptions,
+    bearerToken?: string
+  ) {
+    // Inject authenticated header.
+    this.injectAuthenticatedHeaders(opts, bearerToken);
+    // Inject authenticated request body.
+    if (!bearerToken) {
+      this.injectAuthenticatedRequestBody(opts);
+    }
+  }
+
+  /**
+   * Applies client authentication on the request's header if either
+   * basic authentication or bearer token authentication is selected.
+   *
+   * @param opts The GaxiosOptions whose headers or data are to be modified
+   *   depending on the client authentication mechanism to be used.
+   * @param bearerToken The optional bearer token to use for authentication.
+   *   When this is used, no client authentication credentials are needed.
+   */
+  private injectAuthenticatedHeaders(
+    opts: GaxiosOptions,
+    bearerToken?: string
+  ) {
+    // Bearer token prioritized higher than basic Auth.
+    if (bearerToken) {
+      opts.headers = opts.headers || {};
+      Object.assign(opts.headers, {
+        Authorization: `Bearer ${bearerToken}}`,
+      });
+    } else if (
+      typeof this.clientAuthentication !== 'undefined' &&
+      this.clientAuthentication.confidentialClientType === 'basic'
+    ) {
+      opts.headers = opts.headers || {};
+      const clientId = this.clientAuthentication.clientId;
+      const clientSecret = this.clientAuthentication.clientSecret || '';
+      const base64EncodedCreds = this.crypto.encodeBase64StringUtf8(
+        `${clientId}:${clientSecret}`
+      );
+      Object.assign(opts.headers, {
+        Authorization: `Basic ${base64EncodedCreds}`,
+      });
+    }
+  }
+
+  /**
+   * Applies client authentication on the request's body if request-body
+   * client authentication is selected.
+   *
+   * @param opts The GaxiosOptions whose headers or data are to be modified
+   *   depending on the client authentication mechanism to be used.
+   */
+  private injectAuthenticatedRequestBody(opts: GaxiosOptions) {
+    if (
+      typeof this.clientAuthentication !== 'undefined' &&
```

Comment:
I think we could  use the optional chaining where ever we use this pattern.

Reply:
Done. I agree this is much cleaner and simpler.